### PR TITLE
Convert datetime values to strings before indexing (#3416)

### PR DESCRIPTION
**Pull request by danielsotirhos**
https://github.com/DataBiosphere/azul/issues/3416

Author

- [x] PR title references issue
- [x] PR title matches issue title (preceded by `Fix: ` for bugs)   <sub>or there is a good reason why they're different</sub>
- [x] Title of main commit references issue
- [x] PR is connected to Zenhub issue and description links to issue

Author (reindex)

- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>

Author (freebies & chains)

- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in the chain        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>

Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>
- [x] Added announcement to PR description                  <sub>or this PR does not require announcement</sub>

Author (requirements, before every review)

- [x] Ran `make requirements_update`                        <sub>or this PR leaves requirements*.txt, common.mk and Makefile untouched</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR leaves requirements*.txt untouched</sub>

Author (before every review)

- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

Primary reviewer (after approval)

- [x] Commented in issue about demo expectations            <sub>or labelled issue as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] Moved ticket to Approved column
- [x] Assigned PR to an operator

Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear              <sub>or issue is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to Github
- [x] Branch pushed to Gitlab and added `sandbox` label     <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passed in sandbox                               <sub>or PR is labeled `no sandbox`</sub>
- [x] Started reindex in `sandbox`                          <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                     <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved linked issue to Merged column
- [ ] Pushed merge commit to Github

Operator (after pushing the merge commit)

- [ ] Made announcement requested by author                 <sub>or PR description does not contain an announcement</sub>
- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened the PR chain                                <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate
- [ ] Pushed merge commit to Gitlab                         <sub>or this changes can be pushed later, together with another PR</sub>
- [ ] Deleted PR branch from Github and Gitlab

Operator (reindex) 

- [ ] Started reindex in `dev`                              <sub>or this PR does not require reindexing or does not target `dev`</sub>
- [ ] Checked for failures in `dev`                         <sub>or this PR does not require reindexing or does not target `dev`</sub>
- [ ] Started reindex in `prod`                             <sub>or this PR does not require reindexing or does not target `prod`</sub>
- [ ] Checked for failures in `prod`                        <sub>or this PR does not require reindexing or does not target `prod`</sub>

Operator

- [ ] Unassigned PR

### CODE REVIEW

File: src/azul/indexer/document.py [hunk 1]
```diff
@@ -348,21 +348,22 @@ def from_index(self, value: Number) -> Optional[bool]:
...
     es_type = 'date'
     null = format_dcp2_datetime(datetime(1, 1, 1, tzinfo=timezone.utc))
 
-    def to_index(self, value: Optional[datetime]) -> str:
+    def to_index(self, value: Optional[str]) -> str:
         if value is None:
             return self.null
         else:
-            return format_dcp2_datetime(value)
+            parse_dcp2_datetime(value)  # to verify a valid dcp2 datetime string
```
Comment (achave11) [inline src/azul/indexer/document.py:359]: I am not sure that this comment is necessary. Up to you.

**Review by achave11 — APPROVED** [review]
Looks good. Only made an optional suggestion.

File: src/azul/time.py [hunk 2]
```diff
@@ -126,17 +126,22 @@ def format_dcp2_datetime(d: datetime) -> str:
...
     """
-    require(str(d.tzinfo) == 'UTC', d)
-    date_string = datetime.strftime(d, dcp2_datetime_format)
-    # Work around https://bugs.python.org/issue13305
-    date_string = ('0000' + date_string)[-31:]
-    assert date_string.endswith('+0000'), date_string
-    return date_string[:-5] + 'Z'
+    if d is None:
+        return None
+    else:
```
Comment (hannes-ucsc) [inline src/azul/time.py:138]: A fixpoint like this is a smell. If None is a fixpoint in one function why is it not a fixpoint of every function? 

Also, format and parse functions should be each other's inverse and this breaks that.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: src/azul/plugins/metadata/hca/transform.py [hunk 3]
```diff
@@ -480,18 +483,33 @@ def _visit_file(self, file):
...
         }
 
+    def _entity_date(self, entity: api.Entity):
+        if entity.update_date is None:
+            update_date = None
+        else:
+            update_date = format_dcp2_datetime(entity.update_date)
+        return {
+            'submission_date': format_dcp2_datetime(entity.submission_date),
+            'update_date': update_date,
```
Comment (hannes-ucsc) [inline src/azul/plugins/metadata/hca/transform.py:499]:
```suggestion
            'update_date': None
                           if entity.update_date is None else
                           format_dcp2_datetime(entity.update_date),
```

File: src/azul/time.py [hunk 4]
```diff
@@ -126,6 +126,11 @@ def format_dcp2_datetime(d: datetime) -> str:
     >>> format_dcp2_datetime(datetime(1, 1, 1, tzinfo=timezone.utc))
     '0001-01-01T00:00:00.000000Z'
 
+    >>> format_dcp2_datetime(None)
+    Traceback (most recent call last):
+    ...
+    AttributeError: 'NoneType' object has no attribute 'tzinfo'
+
```
Comment (hannes-ucsc) [inline src/azul/time.py:133]:
```suggestion
```

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Review by hannes-ucsc — APPROVED** [review]

**Comment by jessebrennan** [comment]
Failures during reindexing in dev:
```
[
  {
    "spec": "tdr:broad-jade-dev-data:snapshot/hca_dev_20201203___20210524_lattice:/2",
    "uuid": "40733888-3b9f-500e-84fd-1a18a64aadbe"
  },
  {
    "spec": "tdr:broad-jade-dev-data:snapshot/hca_dev_20201203___20210524_lattice:/2",
    "uuid": "94565940-7b2f-5fd7-beb1-dc309687fe09"
  },
  {
    "spec": "tdr:broad-jade-dev-data:snapshot/hca_dev_20201203___20210524_lattice:/2",
    "uuid": "9d0f5cd1-0ef6-5eed-8dd7-f15d36a34185"
  },
  {
    "spec": "tdr:broad-jade-dev-data:snapshot/hca_dev_20201203___20210524_lattice:/2",
    "uuid": "ae338c4e-650f-545b-8631-3c8d755aa2f9"
  },
  {
    "spec": "tdr:broad-jade-dev-data:snapshot/hca_dev_20201203___20210524_lattice:/2",
    "uuid": "c12a6ca2-3234-5182-ad0b-f8c2fb881ab5"
  }
]
```
These failure are expected because of:
- https://github.com/DataBiosphere/azul/issues/2870
- https://github.com/DataBiosphere/azul/issues/2873

**Comment by hannes-ucsc** [comment]
What's `uuid` above?

**Comment by jessebrennan** [comment]
It's the bundle UUID. To produce this I ran:
```
jq '[.messages[].Body.notification|{"spec": .source.spec, "uuid": .match.bundle_uuid}]' azul-notifications_fail-dev.json 
```

**Pull request merged**